src/nvim/buffer.c

```diff
@@ -3824,7 +3852,143 @@ int build_stl_str_hl(
+          start = item[center_begin].start + fill_left;
+          STRMOVE(start, item[center_begin].start);
+          for (char_u *s = item[center_begin].start; s < start; s++) {
+            *s = fillchar;
+          }
+        } else {
+          // Handle any expected concatenation problems
+          start = item[middle_mark].start + fill_left;
+          for (char_u *s = item[center_begin].start; s < start; s++) {
+            if (s == out + len_a) {
+              *s = '<';
+            } else {
+              *s = *(s - fill_left);
+            }
+          }
+        }
+
+        int item_idx = center_begin;
+        // Adjust the offset of any items after the middle
+        for (item_idx++; item_idx < itemcnt; item_idx++) {
+          item[item_idx].start += fill_left;
+        }
+
+        // Get the new lengths that were specified above
+        int len_d = STRLEN(out);
+
+        int fill_right = maxwidth - len_d;
+
+        char_u *end;
+        if (fill_right >= 0) {
+          // This is a non truncating section
+          end = item[center_end].start + fill_right;
+          STRMOVE(end, item[center_end].start);
+          if (fill_left < 0) {
+            end = item[center_end].start + fill_right - fill_left;
+          }
+          for (char_u *s = item[center_end].start; s < end; s++) {
+            *s = fillchar;
+          }
+        } else {
+          // Handle the truncation here
+          end = out + STRLEN(out) + fill_right;
+          for (char_u *s = item[middle_mark].start; s < end; s++) {
+            if (s == item[middle_mark].start) {
+              *s = '<';
+            } else {
+              *s = *(s - fill_right);
+            }
+          }
+        }
+
+        width = maxwidth;
+      } else {
+        // Center the text within center_begin and center_end
+        int center_begin_fill = (maxwidth - width) / 2;
+        int center_end_fill = (maxwidth - width) - center_begin_fill;
+
+        // TODO(tjdevries): Better naming for some of these variables
+        char_u *start = item[center_begin].start + center_begin_fill;
+        STRMOVE(start, item[center_begin].start);
+
+        // Fill the <text>[<text> to be <text>     <text>
+        for (char_u *s = item[center_begin].start; s < start; s++) {
+          *s = fillchar;
+        }
+
+        int item_idx = center_begin;
+        // Adjust the offset of any items after the middle
+        for (item_idx++; item_idx <= center_end; item_idx++) {
+          item[item_idx].start += center_begin_fill;
+        }
+
+        char_u *end = item[center_end].start + center_end_fill;
+        STRMOVE(end, item[center_end].start);
+
+        // Fill the <text>]<text to be <text>      <text>
+        for (char_u *s = item[center_end].start; s < end; s++) {
+          *s = fillchar;
+        }
+
+        width = maxwidth;
+      }
+    // Otherwise, we just do the exact same thing with the middle that
+    //  we used to do.
+    } else {
```

Comment:
Okay, that's nice to hear. I will go back and split a lot of this functionality out into more manageable functions. I can just ping both of you ( @jamessan & @justinmk ) once that's completed? It may take me a bit because I'd also like to add some unit tests for those functions if that's alright.